### PR TITLE
Resolvers Support

### PR DESCRIPTION
**Pull request by volodymyr-mykhailyk**
Support setting value resolving using custom backends

### CODE REVIEW

**Review by volkorny — APPROVED** [review]

File: lib/consul_application_settings/utils.rb [hunk 1]
```diff
@@ -34,6 +36,16 @@ def cast_complex_value(value)
...
         value.to_s
       end
+
+      def convert_to_hash(data)
+        data_h = data.map do |item|
+          value = cast_consul_value(item[:value])
+          item[:key].split('/').reverse.reduce(value) { |h, v| { v => h } }
+        end
+        data_h.reduce({}) do |dest, source|
+          DeepMerge.deep_merge!(source, dest, preserve_unmergeables: true)
```
Comment (vbyno) [inline lib/consul_application_settings/utils.rb:46]: as far as I remember, modifying inside `reduce` works slowly. Maybe, replace `reduce` with `each_with_object`, or use `deep_merge` (with no exclamation mark)

**Review by vbyno — APPROVED** [review]

File: lib/consul_application_settings/resolvers/abstract.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+module ConsulApplicationSettings
+  module Resolvers
+    # Abstract resolver with basic functionality
+    class Abstract
```
[the remark this answers is not in the document]
Comment (vbyno) [reply lib/consul_application_settings/resolvers/abstract.rb:4]: oh, I see, it is used it in specs

File: spec/consul_application_settings/reader_spec.rb [hunk 3]
```diff
@@ -85,4 +90,41 @@
       end
     end
   end
+
+  context 'custom resolvers' do
```
Comment (danpolyuha) [inline spec/consul_application_settings/reader_spec.rb:94]: `context` -> `describe`

File: spec/consul_application_settings/reader_spec.rb [hunk 4]
```diff
@@ -85,4 +90,41 @@
...
+    describe '#get' do
+      it 'resolves value from local config' do
+        expect(reader.get('application/name')).to eq('resolved')
+      end
+
+      it 'resolves value from consul' do
+        expect(reader.get('application/key')).to eq('resolved')
+      end
+
+      it 'return original if not resolvable' do
```
Comment (danpolyuha) [inline spec/consul_application_settings/reader_spec.rb:120]: `returnS`

File: spec/consul_application_settings/resolvers/abstract_spec.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+RSpec.describe ConsulApplicationSettings::Resolvers::Abstract do
+  subject(:resolver) { described_class.new }
+  let(:value) { [true, 'env://eeee', 1, nil].sample }
+  let(:path) { 'test/call' }
+
+  it 'do not resolve value any value' do
```
Comment (danpolyuha) [inline spec/consul_application_settings/resolvers/abstract_spec.rb:6]: `does not resolve to any value`

File: spec/consul_application_settings/resolvers/erb_spec.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    it 'has access to env' do
+      expect(resolver.resolve("<%= ENV['PATH'] %>", path)).to eq(ENV['PATH'])
+    end
+
+    context 'with invalid input' do
+      it 'ignores incorrect erb' do
+        expect(resolver.resolve('aaaa<%= 2 +', path)).to eq('aaaa 2 +')
+      end
+
+      it 'raising error for missing variable' do
```
Comment (danpolyuha) [inline spec/consul_application_settings/resolvers/erb_spec.rb:41]: `raises`

**Review by danpolyuha — APPROVED** [review]
I like the idea and everything very much.
Really hope we'll be able to replace `config` gem on all projects soon.

File: lib/consul_application_settings/providers/consul_preloaded.rb [hunk 7]
```diff
@@ -4,24 +4,42 @@ module Providers
...
       end
+
+      def resolve_tree_response(value)
+        return value unless value.is_a?(Hash)
+
+        deep_cast(value)
+      end
+
+      def deep_cast(hash)
+        hash.each_pair do |k, v|
```
[the remark this answers is not in the document]
Comment (volodymyr-mykhailyk) [reply lib/consul_application_settings/providers/consul_preloaded.rb:35]: That's how Diplomat returns value via `recurse: true`

File: lib/consul_application_settings/resolvers/env.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
+module ConsulApplicationSettings
+  module Resolvers
+    # Resolve values in environment
+    class Env
+      IDENTIFIER = 'env://'.freeze
+
+      def resolvable?(value, _path)
+        return unless value.respond_to?(:start_with?)
```
[the remark this answers is not in the document]
Comment (volodymyr-mykhailyk) [reply lib/consul_application_settings/resolvers/env.rb:8]: yeap. Will cleanup in future update

File: spec/consul_application_settings/resolvers/abstract_spec.rb [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+    expect(resolver.resolvable?(value, path)).to be_falsey
```
[the remark this answers is not in the document]
Comment (volodymyr-mykhailyk) [reply spec/consul_application_settings/resolvers/abstract_spec.rb:7]: make sense. I will also update it

File: .rubocop.yml [hunk 10]
```diff
@@ -10,4 +11,4 @@ Style/FrozenStringLiteralComment:
   Enabled: false
 
 Metrics/BlockLength:
-  ExcludedMethods: ['describe', 'context']
\ No newline at end of file
+  ExcludedMethods: ['describe', 'context', 'shared_examples']
```
[the remark this answers is not in the document]
Comment (volodymyr-mykhailyk) [reply .rubocop.yml:14]: Funny that rubocop does not validate it's own file :)

File: lib/consul_application_settings/reader.rb [hunk 11]
```diff
@@ -21,5 +19,32 @@ def load(sub_path)
...
       self.class.new(new_path, @config)
     end
+
+    protected
+
+    def check_deep_structure(value, path)
+      return unless value.is_a?(Hash)
+
+      message = "Getting value of complex object at path: '#{path}'. Use #load method to get new scoped instance"
+      raise ConsulApplicationSettings::Error, message if value.is_a?(Hash)
```
[the remark this answers is not in the document]
Comment (volodymyr-mykhailyk) [reply lib/consul_application_settings/reader.rb:29]: Yes. forgot to remove after refactoring it

File: lib/consul_application_settings/resolvers/abstract.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (volodymyr-mykhailyk) [reply lib/consul_application_settings/resolvers/abstract.rb:4]: Wanted to have documented interface if anyone is going to implement their own resolver

File: lib/consul_application_settings/resolvers/env.rb [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+        value.start_with?(IDENTIFIER)
+      end
+
+      def resolve(value, _path)
+        env_path = value.to_s.delete_prefix(IDENTIFIER)
+        ENV[env_path]
```
[the remark this answers is not in the document]
Comment (volodymyr-mykhailyk) [reply lib/consul_application_settings/resolvers/env.rb:15]: I'm not sure. We are not raising errors when the setting path is incorrect. Why should we be doing it here?

**Comment by volodymyr-mykhailyk** [comment]
Thanks for the comments. I will fix typos, suggestion in following Pull request

**Pull request merged**